### ISSUE TITLE
[Table] In-line editing of Data Table 

### ISSUE THREAD

**Issue by davidicus** [IBM/carbon-addons-iot-react#720]
**Description**
Extend the IoT Data Table to allow for editing of lines row-by-row

**Name of Pattern or Asset**
Data Table

**User outcome**
Use can edit the data of visible columns using appropriate component, and save multiple row edits at once

**Proposal**
https://jsw.ibm.com/browse/GRAPHITE-3571
![screen shot 2019-10-29 at 4 48 13 pm](https://media.github.ibm.com/user/17018/files/f3308a00-fa6b-11e9-81bd-a11205e73072)
<img width="634" alt="screen shot 2019-11-14 at 9 41 18 am" src="https://media.github.ibm.com/user/17018/files/f316c800-06c2-11ea-9200-1a6db6fcab6b">

**Additional context**
https://jsw.ibm.com/browse/GRAPHITE-3571
- Add an icon to toggle in-line edit mode on in the top-right of the data table
-  Replace other content in the table header with Save and Cancel buttons at top when edit mode is on
-  Programmatically display the correct component for editing the fields shown, depending on the data source. Date fields are displayed with date picker, numeric with numeric spinner, alphanumeric with text entry, etc.  
    - Issue to add Date picker and Numeric spinner to design library: https://github.ibm.com/Watson-IoT/pal-tracking/issues/6
- Components should display with proper padding and alignment within the width of the row/column
- Fields should be validated on entry, before save (Install date cannot be in the future, End of Life cannot be in the past, Expected life must be greater than or equal to 1, etc) 
    - Issue for error state: https://github.ibm.com/Watson-IoT/pal-tracking/issues/32
- Display toast with undo option https://ibm.invisionapp.com/d/main#/console/15350624/319220019/preview (Use light version of toast. Dark version does not match PAL update)

**Comment by davidicus** [comment]
PAL tracking issue: https://github.ibm.com/Watson-IoT/pal-tracking/issues/20
Assigned dev: https://github.ibm.com/lijianig
waiting on dev to communicate their public github username

**Comment by bjornalm** [comment]
The spec shows this: 
![image](https://user-images.githubusercontent.com/5167091/80114844-536c3880-8584-11ea-97b2-e959d87ae71d.png)

But we already have a delete/cancel design for multiple rows which is this one:
![image](https://user-images.githubusercontent.com/5167091/80114961-78f94200-8584-11ea-82f3-4cab5cb1c5d3.png)

Would it not make sense to use the same design and code for the row editing (but with save/cancel buttons instead)?

**Comment by davidicus** [comment]
@lukefirth I agree this sounds ideal. What are your thoughts?

**Comment by davidicus** [comment]
@JordanWSmith15 @KarenGski was this option explored at all? what are your thoughts?

**Comment by JordanWSmith15** [comment]
I have no preference from a design perspective. We should not show the "X items selected" string if we go with the blue bar.

**Comment by KarenGski** [comment]
I'm fine using the same mechanism for both multi-select and inline edit. 

As Jordan said, the items selected isn't relevant because you aren't doing bulk actions....  unless there IS a use case to allow the user to perform batch inline edits.  We don't currently have it, but I could imagine a user wanting to select a few things and then change all the values at once.  Maybe a future enhancement.

**Comment by lukefirth** [comment]
I think the challenge/difference here is that the bulk edit of the data table is wholesale (it's for every single row) whereas the bulk action bar **only** deals with performing actions on the currently selected (denoted by the checkboxes) rows, as stated above.

In other words, I believe we'd run into conflicts with a user editing many rows, and then having the action bar/checkboxes have an uncertain relationship.

We _could_ make this work, by removing the checkboxes and the X' items selected' when in an edit mode, making it seem like 'all rows' are selected implicitly when in bulk edit mode - which would allow us to use the action bar styling, but that does sound A: A bit more complicated than I'd like (hacking both the main table and the action bar a bit), and B: Reliant on users understanding the implict 'this action bar is referring to any changes I made anywhere in the cells) in order to work. 🤔 hmmm.

**Comment by davidicus** [comment]
@lukefirth I hear what you are saying and it makes sense. FWIW, in my personal perspective I view the bulk action bar to just be about bulk actions and not explicitly tied to checkboxes. It feels like two separate things. One is a mechanism for selecting what fields you want to to have action applied to (checkboxes/ edit button) and then the action bar which is just the action you are going to apply to whatever is selected. It does seem weird to have two different ways to apply bulk action. Totally understand that others may not draw those same conclusions.

**Comment by bjornalm** [comment]
Thanks everyone for pitching in. To clarify my proposal, there would be no checkboxes and no "X items selected" in the case of row editing. David described the separation of concerns well above. It could look like this:

![image](https://user-images.githubusercontent.com/5167091/80188307-42b5d400-8611-11ea-8085-28f57bd74180.png)

**Comment by lukefirth** [comment]
I do get what you're saying (so thats good, we at least all understand each others perspectives ha).

I *do* like the primacy that the bar gives to the save/action - it makes it very clear the table is in a temporary state and an action 'there' is needed to release it back to its regular state. This is nice and appropriate. 

However, I do still reach a kind of discomfort in using the action bar styling for this. Which I'll try to ellucidate:
 
- In the grand scheme of patterns outside of the data table, this may be considered a very special kind of 'form', kind of in the same way that the dashboard editor can be seen as one too. In all cases where there is a save/cancel pair, we both have a hard pattern of 'save on the right, cancel on the left' and a styling choice of primary for save, secondary for cancel as a button pair. Using this action bar styling while matching other 'bulk actions' in the table. Long way of saying - this may make it partially more inline with other things in the sphere of the table, but would break it further from any other cancel/save patterns in our applications.

- Normally, the actions in the bar are always initiating something (or the 'null' action of dismissing the action bar) - and so can adopt a primary styling. However, this state is no longer showing a pallette of initiate-able actions but a 'positive' and a 'negative' action (to cancel). Which makes it a little different in that it reuses the 'cancel' of the action bar in a slightly different way than it is regularly used. In 'normal' cases the action bars cancel action would be a null type action. _"I started going to make a bulk action, but decided not to so I clicked the cancel and it just deselected things'_ - nothing is lost. However here the cancel is used for a negative action. Clicking it will likely be a destructive act - losing any changes you've already made. This one is a really finicky thing I know, so treat this point as a lesser one.

- This introduces a second action bar (you could still in multiselect tables have checkboxes that will trigger the action bar) which I'm not sure is actually a good thing. They will look very similar, but intiated in different ways, have different content and behave in the different ways outlined above. 

Really point 1 is my main concern from a UX perspective - making it more like regular bulk actions (when it is actually a seperate action bar), and departing it from the standard save/cancel button pair we use for any other kind of edit/create states. 

However, I'd love to get a visual design expert eye on this too. @amymillett , would you mind pitching in?

**Comment by bjornalm** [comment]
Since the row actions can conflict with the edit mode (think deleting a row while a cell is edited) we should probably also disable the row actions. Also, the row actions as they currently are implemented could obscure cells on row hover, preventing the user from clicking/editing the cell content using the mouse.

**Comment by KarenGski** [comment]
Agree with Luke that the multi-select boxes would have to be disabled for this to make sense at at this point. 
But again.. there could be a case in the future where you'd want to select more than one row and then click into a field on any of the rows and have the change for that field apply to all of them.  I've seen that in things like RTC before but not sure how much people need it.

**Comment by bjornalm** [comment]
So @lukefirth, what do I do? 
1. Should I go with white the cancel/save bar like in the original proposal or with the blue bar?
2. Should I also completely disable the row actions like I suggest above?

**Comment by lukefirth** [comment]
1. Go with the original proposal.
2. The challenge there is that disabling them completely will also stop users being able to remove a row, which is a normal part of most editing cases. While I agree with your point, I do not think completely disabling will work for this reason, unless there is another mechanism a team can use to ensure a remove action would be able to be present in the row when changed to edit mode without it being a regular action.

**Comment by bjornalm** [comment]
@stuckless See the discussion about row actions above, are you ok with allowing them for now and have Luke define a better solution for them in a separate issue?

**Comment by bjornalm** [comment]
@lukefirth The default buttons are slightly wider and don't center the text, is that ok or should I modify that so that they look exactly like in the design spec?

Default buttons:
![image](https://user-images.githubusercontent.com/5167091/80635375-17d9de80-8a5c-11ea-93fb-174e9347bb77.png)

**Comment by stuckless** [comment]
@bjornalm I'm ok with the row actions.  We can always disable them ourselves, or completely remove them, by hooking into the table's edit state.

**Comment by bjornalm** [comment]
@stuckless @lukefirth Do we need row editing in the TableCard as well?

**Comment by lukefirth** [comment]
@bjornalm 

- **Button style:** Apologies for this, we're still working out some design 'kinks'. Use the default style, but please switch the order to place the save on the right, cancel on the left (I thought I'd caught all instances of it the other way around...)

- **Row edit in table cards** I have no tracked requirement for this at present from the teams, but it does seem like something that would be only a matter of time before one of our teams using table cards ask for it. My recommendation here is that if its easy to do now, do so, but if its any real work at all to make it work in table cards - then we should leave it and hit whatever our next priority is and return if/when it comes up as a priority.
